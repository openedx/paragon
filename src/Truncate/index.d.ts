import * as React from 'react';

export interface TruncateProps {
  children: string;
  lines?: string | number;
  ellipsis?: string | number | React.ReactNode;
  whiteSpace?: boolean;
  elementType?: string;
  className?: string;
  onTruncate?: (truncated: HTMLElement[]) => void;
}

declare const Truncate: React.FC<TruncateProps>;

export default Truncate;
