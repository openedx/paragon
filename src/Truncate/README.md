---
title: 'Truncate'
type: 'component'
components:
- Truncate
status: 'New'
designStatus: 'Done'
devStatus: 'Done'
---

A Truncate component can help you crop multiline text. There will be three dots at the end of the text.

## Basic Usage

```jsx live
  <Truncate lines={2}>
    Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons 
    for using the platform and objectives to accomplish. To help members of each group learn about what edX 
    offers, reach goals, and solve problems, edX provides a variety of information resources.
    Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons
    for using the platform and objectives to accomplish. To help members of each group learn about what edX
    offers, reach goals, and solve problems, edX provides a variety of information resources.
  </Truncate>
```

### With the custom ellipsis

```jsx live
    <Truncate lines={2} ellipsis="ðŸŽ‰ðŸŽ‰ðŸŽ‰" whiteSpace>
        Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons
        for using the platform and objectives to accomplish. To help members of each group learn about what edX
        offers, reach goals, and solve problems, edX provides a variety of information resources.
    </Truncate>
```

### With the onTruncate

```jsx live
    <Truncate lines={2} onTruncate={() => console.log('onTruncate')}>
        Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons
        for using the platform and objectives to accomplish. To help members of each group learn about what edX
        offers, reach goals, and solve problems, edX provides a variety of information resources.
    </Truncate>
```

### Example usage in Card

```jsx live
() => {
  const isExtraSmall = useMediaQuery({ maxWidth: breakpoints.extraSmall.maxWidth });
  
  return (
    <Card style={{ width: isExtraSmall ? "100%" : "18rem" }} isClickable>
      <Card.ImageCap
        src="https://picsum.photos/360/200/"
        srcAlt="Card image"
      />
      <Card.Header
        title={
          <Truncate lines={2}>
            Using Enhanced Capabilities In Your Course
          </Truncate>}
      />
      <Card.Section>
        <Truncate lines={4}>
          Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons
          for using the platform and objectives to accomplish. To help members of each group learn about what edX
          offers, reach goals, and solve problems, edX provides a variety of information resources.
        </Truncate>
      </Card.Section>
      <Card.Footer
        textElement={
          <Truncate lines={2}>
            Using Enhanced Capabilities In Your Course
          </Truncate>}
      >
        <Button style={{ minWidth: 100 }}>Action 1</Button>
      </Card.Footer>
    </Card>
  )
}
```

### HTML markdown support

**Note**: `Truncate` supports only plain `HTML` children and not `jsx`.

```jsx live
<Truncate lines={1}>
  <a href="#">Learners</a>, course teams, researchers, developers: the edX community includes <strong class="strong-class"><i class="i-class">groups with <u>a range</u> of <q>reasons</q></i></strong> for using the platform and objectives to accomplish.
</Truncate>
```
