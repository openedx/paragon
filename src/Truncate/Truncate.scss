// Put styles related to your component here

@import "variables";
