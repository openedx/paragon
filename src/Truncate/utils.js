const createTextClamp = (text, coefficient) => {
  const startIndex = 0;
  const sliceIndex = Math.floor(text.length * coefficient);
  return text.slice(startIndex, sliceIndex);
};

const createCopyElement = (sourceElement, sourceElementStyles) => {
  const newElement = document.createElement(sourceElement.tagName);
  const sourceStyles = window.getComputedStyle(sourceElement);
  console.log('sourceElement', sourceElement);
  newElement.style.cssText = sourceStyles.cssText;
  newElement.style.maxWidth = sourceStyles.width;

  Object.keys(sourceElementStyles).forEach(property => {
    newElement.style[property] = sourceElementStyles[property];
  });

  return newElement;
};

const constructString = (string, whiteSpace, ellipsis) => {
  const spacer = whiteSpace ? ' ' : '';
  return `${string.trim()}${spacer}${ellipsis}`;
};

const clampLines = (text, element, { lines, whiteSpace, ellipsis }) => {
  const lineHeightValue = 200;
  const maxHeight = lineHeightValue * Number(lines);
  const maxWidth = 1500;
  const DEFAULT_PADDING_VALUE = 0;
  const ellipsisElement = createCopyElement(element, {
    lineHeight: `${lineHeightValue}px`,
    height: 'auto',
    position: 'absolute',
    opacity: '0',
    left: '-1px',
    width: `${element.scrollWidth}px`,
    paddingTop: DEFAULT_PADDING_VALUE,
    paddingBottom: DEFAULT_PADDING_VALUE,
  });

  element.appendChild(ellipsisElement);

  let clampedText = text;
  ellipsisElement.innerHTML = constructString(clampedText, whiteSpace, ellipsis);

  let ellipsisElementHeight = Math.ceil(ellipsisElement.scrollHeight);
  if (ellipsisElementHeight <= maxHeight) {
    ellipsisElement.parentNode.removeChild(ellipsisElement);

    return clampedText;
  }

  let decrementCoefficient = (maxHeight / ellipsisElementHeight) + 0.35;

  while (ellipsisElementHeight > maxHeight && clampedText.length) {
    clampedText = createTextClamp(text, decrementCoefficient);
    ellipsisElement.innerHTML = constructString(clampedText, whiteSpace, ellipsis);
    ellipsisElementHeight = Math.ceil(ellipsisElement.scrollHeight) - 1;
    decrementCoefficient -= text.length > maxWidth ? 0.011 : 0.018;
  }

  ellipsisElement.parentNode.removeChild(ellipsisElement);
  return constructString(clampedText, whiteSpace, ellipsis);
};

module.exports = {
  createTextClamp,
  clampLines,
  createCopyElement,
  constructString,
};
