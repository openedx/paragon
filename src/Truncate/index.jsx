import React, {
  useLayoutEffect, useRef,
} from 'react';
import PropTypes from 'prop-types';
import { truncateLines } from './utils';
import useWindowSize from '../hooks/useWindowSize';

const DEFAULT_TRUNCATE_LINES = 1;
const DEFAULT_TRUNCATE_ELLIPSIS = '...';
const DEFAULT_TRUNCATE_ELEMENT_TYPE = 'div';

function Truncate({
  children, lines, ellipsis, elementType, className, whiteSpace, onTruncate,
}) {
  const textContainer = useRef();
  const { width } = useWindowSize();

  useLayoutEffect(() => {
    if (textContainer.current) {
      const [truncated, original] = truncateLines(children, textContainer.current, {
        ellipsis,
        whiteSpace,
        lines,
      });
      textContainer.current.setAttribute('title', original);
      textContainer.current.setAttribute('aria-label', original);
      textContainer.current.innerHTML = '';
      truncated.forEach(element => {
        textContainer.current.appendChild(element);
      });
      if (onTruncate) {
        onTruncate(truncated);
      }
    }
  }, [children, ellipsis, lines, onTruncate, whiteSpace, width]);

  return React.createElement(elementType, {
    ref: textContainer,
    className,
  });
}

Truncate.propTypes = {
  /** The expected text to which the ellipsis would be applied. */
  children: PropTypes.string.isRequired,
  /** The number of lines the text to be truncated to. */
  lines: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Text content for the ellipsis - will appear after the truncated lines. */
  ellipsis: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.node]),
  /** Adds the whitespace from before the ellipsis. */
  whiteSpace: PropTypes.bool,
  /** Custom html element for truncated text. */
  elementType: PropTypes.string,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Callback fired when a text truncating */
  onTruncate: PropTypes.func,
};

Truncate.defaultProps = {
  lines: DEFAULT_TRUNCATE_LINES,
  ellipsis: DEFAULT_TRUNCATE_ELLIPSIS,
  whiteSpace: false,
  elementType: DEFAULT_TRUNCATE_ELEMENT_TYPE,
  className: undefined,
  onTruncate: undefined,
};

export default Truncate;
