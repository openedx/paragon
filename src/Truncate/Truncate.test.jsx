import React from 'react';
import { mount } from 'enzyme';
import Truncate from './index';

describe('<Truncate />', () => {
  const wrapper = mount(
    <Truncate>
      Learners, course teams, researchers, developers.
    </Truncate>,
  );
  it('render with className', () => {
    wrapper.setProps({ className: 'pgn__truncate' });
    expect(wrapper.hasClass('pgn__truncate')).toEqual(true);
  });
  it('render with onTruncate', () => {
    const mockFn = jest.fn();
    wrapper.setProps({ onTruncate: mockFn });
    expect(mockFn.mock.calls.length).toBe(1);
  });
});
