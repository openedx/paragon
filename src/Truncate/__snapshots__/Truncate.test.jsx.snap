// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Truncate /> Component successfully renders 1`] = `
<div
  className={null}
>
  Learners, course teams, researchers, developers: the edX community includes groups with a range of reasons for using the platform and objectives to accomplish. To help members of each group learn about what edX offers, reach goals, and solve problems, edX provides a variety of information resources.
</div>
`;
